synapse/storage/databases/state/store.py

```diff
@@ -202,7 +202,14 @@ def _get_state_for_group_using_cache(
                 requests state from the cache, if False we need to query the DB for the
                 missing state.
         """
-        cache_entry = cache.get(group)
+        # If we are asked explicitly for a subset of keys, we only ask for those
+        # from the cache. This ensures that the `DictionaryCache` can make
+        # better decisions about what to cache and what to expire.
+        dict_keys = None
+        if not state_filter.has_wildcards():
+            dict_keys = state_filter.concrete_types()
+
+        cache_entry = cache.get(group, dict_keys=dict_keys)
```

Comment:
ideally we'd separate this to a different PR. It's not directly related to the changes to DictionaryCache.